### PR TITLE
[FLINK-21768][clients]Optimize system.exit() logic of CliFrontend

### PR DESCRIPTION
**Pull request by zuston**
Change-Id: Ia28bb5b7d389b3faedb11af15576a458a74d4e76

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 1]
```diff
@@ -1124,20 +1124,21 @@ public static void main(final String[] args) {
...
-            int retCode =
+            retCode =
                     SecurityUtils.getInstalledContext().runSecured(() -> cli.parseAndRun(args));
-            System.exit(retCode);
         } catch (Throwable t) {
             final Throwable strippedThrowable =
                     ExceptionUtils.stripException(t, UndeclaredThrowableException.class);
             LOG.error("Fatal error while running command line interface.", strippedThrowable);
             strippedThrowable.printStackTrace();
             System.exit(31);
```
Comment (SteNicholas) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:1139]: retCode = 31, and call `System#exit` method in finally block.

Comment (zuston) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: Done.

**Comment by zuston** [comment]
@SteNicholas Please review again. Thanks.

**Comment by SteNicholas** [comment]
@wangyang0918 Please help @zuston to review this pull request.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 2]
```diff
@@ -1124,19 +1124,19 @@ public static void main(final String[] args) {
...
                 loadCustomCommandLines(configuration, configurationDirectory);
 
+        int retCode = 31;
         try {
             final CliFrontend cli = new CliFrontend(configuration, customCommandLines);
 
             SecurityUtils.install(new SecurityConfiguration(cli.configuration));
-            int retCode =
-                    SecurityUtils.getInstalledContext().runSecured(() -> cli.parseAndRun(args));
-            System.exit(retCode);
```
Comment (wangyang0918) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:1133]: Do you mean the Oozie will handle the system exit signal and throw an exception here? After then, we will go into the `catch` code blocks and call the system exit again.

So you are suggesting to ensure that only call the `System.exit` once. Right?

Comment (zuston) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:1133]: Yes.

**Comment by zuston** [comment]
Any update on it? @wangyang0918

**Review by wangyang0918 — APPROVED** [review]
Changes look good to me. +1 for merging.

**Review by SteNicholas — APPROVED** [review]

**Comment by wangyang0918** [comment]
cc @tillrohrmann would you mind to have a final check and help with merging?

**Comment by wangyang0918** [comment]
@zuston nit: The commit message could be refined.
*  Commit title`[clients]` -> `[client]`
* Remove the magic commit message content `Change-Id: Ia28bb5b7d389b3faedb11af15576a458a74d4e76`

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @zuston. LGTM. Merging once AZP gives green light.

**Comment by zuston** [comment]
> @zuston nit: The commit message could be refined.
> 
> * Commit title`[clients]` -> `[client]`
> * Remove the magic commit message content `Change-Id: Ia28bb5b7d389b3faedb11af15576a458a74d4e76`

Done. 
I share project with internal gerrit, so it will add change-id with gerrit.

**Comment by zuston** [comment]
> Thanks for creating this PR @zuston. LGTM. Merging once AZP gives green light.

Thanks.

**Comment by tillrohrmann** [comment]
`CoordinatedSourceRescaleITCase.testUpscaling` failed in the latest build. @zuston could you check whether this is related test failure? If not, then let's open a JIRA for it.

**Comment by tillrohrmann** [comment]
Actually it is unrelated. I will open the ticket.

**Comment by zuston** [comment]
> Actually it is unrelated. I will open the ticket.

Thanks. @tillrohrmann 
Besides, i have a unrelated with this PR question that why many PR is closed although Pull Request's code have been merged into Flink master branch?
Why not we directly use github PR merge action?  Because It makes confused when I surf the Flink github PR list.

**Comment by tillrohrmann** [comment]
@zuston this is a bit of an artifact of our infrastructure. Not every committer commits via the Github web UI but many (myself included) prefer the git CLI. If people use the git CLI, then there is a bot which reads the latest commits and looks for a `This closes #PR-Number.` line. If it finds this line, then it knows that it can close the PR with `PR-Number`. That's why PRs get closed instead "merged". In the end the code will be in the master branch.

**Comment by zuston** [comment]
Thanks for review and merge @tillrohrmann @wangyang0918 @SteNicholas

**Pull request closed without merging**